docs/proposal/20241003-data-disk-support.md

```diff
@@ -0,0 +1,170 @@
+type VSphereDisk struct {
+	// SizeGiB is the size of the disk (in GiB).
+	// +kubebuilder:validation:Required
+	SizeGiB int32 `json:"sizeGiB"`
+}
+```
+
+The above type has been added to the machine spec section
+
+```go
+// VirtualMachineCloneSpec is information used to clone a virtual machine.
+type VirtualMachineCloneSpec struct {
+    ...
+    // DataDisks holds information for additional disks to add to the VM that are not part of the VM's OVA template.
+    // +optional
+    DataDisks []VSphereDisk `json:"dataDisks,omitempty"`
+}
+```
+
+### Implementation Details
+
+The cloning process will now be able to add data disks to a VM.  Using the config provided in the VSphereCloneTemplate, the defined disks will be added to the virtual machine.  Each disk is added to the controller used by the primary disk.  Currently, there are no plans to allow the user to define a new controller (scsi/ide/etc) for these disks. 
+
+An example of what the VSphereMachineTemplate looks like when data disks are desired:
+```yaml
+apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
+kind: VSphereMachineTemplate 
+metadata:
+  name: test-machine-template
+  namespace: openshift-cluster-api
+spec:
+  template:
+    spec: 
+      template: CAPV_2_Disk
+      server: vcenter.example.com
+      diskGiB: 128
+      dataDisks:
+      - sizeGiB: 50
+      - sizeGiB: 90
+      cloneMode: linkedClone
+      datacenter: cidatacenter
+      datastore: /cidatacenter/datastore/vsanDatastore
+      folder: /cidatacenter/vm/multi-disk-k96l6
+      resourcePool: /cidatacenter/host/cicluster
+      numCPUs: 4
+      memoryMiB: 16384
+      network:
+        devices:
+        - dhcp4: true
+          networkName: ci-vlan-1240
+```
+
+In the above example, two data disks will be created during the clone process and placed after the OS disk.  The example shows a 50 GiB and a 90 GiB disk being configured and added.
+
+For each `dataDisks` definition, the clone procedure will attempt to generate a device VirtualDeviceConfigSpec that will be used to create the new disk device.  Each disk will be attached in the order in which they are defined in the template.  All disks defined in the vSphere OVA template will come first with the new disks being attached after.
+
+The clone procedure will assign each new disk to the same controller being used by the primary (OS) disk of the OVA template.  The current behavior of the clone procedure will not be able to create any new controllers; however, in the future, the template may be enhanced to define new controllers and then assign them during the disk creation portion of the clone procedure. 
+
+### Notes/Constraints
+
+## Upgrade Strategy
+
+No Upgrade needed.
+Data disks are optional. Existing clusters do not have additional data disks configured, which will continue to work without changes.  Once upgraded, the new feature will be available for existing machines to be reconfigured / recreate.
```

Comment:
if the CAPV template doesn't have dataDisks set, but it's added later would the webhook block the edit?
i.e. one needs to create a new template.

Reply:
I was just starting to look at that and forgot to revisit.  In OCP, I have our MAPI operator allowing the change to a machineset which then requires the admin to roll out the changes manually (either delete the machine they wish to remove first, or scale down then up).  I'll look back at this and add a section on this here.